### PR TITLE
WIP: Substrate wasm processor

### PR DESCRIPTION
**Pull request by jiqiang90**
TODO:

- [x] wasm event processor 
- [x] wasm call processor
- [x] Update subql/types `filterProcessor` to take `assets`， also need update in subql/node
- [x] contract provider
- [ ] dynamic datasource
- [ ] setup dictionary
- [x] example project 
- [ ] and documentation

Currently running into issue with wasm-starter project, but unit test with package decoding works fine.

```
 Error: createType(Option<AccountId>):: Unknown type passed to AccountId constructor, found typeof object

```

### CODE REVIEW

**Comment by ianhe8x** [comment]
Don't call it astar wasm, use substrate wasm instead

File: packages/substrate-wasm/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+  //This should align with https://github.com/polkadot-js/api/blob/0b6f7861080c920407a346e2a3dbe64adcb07a1e/packages/api-contract/src/Abi/index.ts#L249
+  const [, trimmed] = compactStripLength(data);
+  return u8aToHex(trimmed.subarray(0, 4));
+}
+
+const EventProcessor: SecondLayerHandlerProcessor_1_0_0<
+  SubstrateHandlerKind.Event,
+  WasmEventFilter,
+  WasmEvent,
+  AnyTuple, // TODO improve this type
```
Comment (stwiname) [inline packages/substrate-wasm/src/index.ts:282]: I think this can be `ContractEmittedResult`

File: packages/substrate-wasm/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+    }
+    return queryEntry;
+  },
+};
+
+const CallProcessor: SecondLayerHandlerProcessor_1_0_0<
+  SubstrateHandlerKind.Call,
+  WasmCallFilter,
+  WasmCall,
+  AnyTuple, // TODO improve this type
```
Comment (stwiname) [inline packages/substrate-wasm/src/index.ts:383]: Can we provide this type now?

File: packages/substrate-wasm/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+      blockHash: original.block.block.hash.toHex(), // Substrate block hash
+      timestamp: original.block.timestamp,
+    };
+    return [call];
+  },
+  async filterProcessor({ds, filter, input}): Promise<boolean> {
+    try {
+      const [dest, , , , data] = input.extrinsic.args;
+      const iAbi = await buildAbi(ds);
+      const decodeMessage = iAbi?.decodeMessage(data.toU8a());
```
Comment (stwiname) [inline packages/substrate-wasm/src/index.ts:428]: Could this be moved below the contract address check? Could speed things up if decoding is slow

File: packages/substrate-wasm/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+  method?: string; // label
+}
+
+const dsAssets: Record<string, string> = {};
+// used abi name with its ABI object will be stored
+const contractAbis: Record<string, Abi> = {};
+//only single event will be stored, reset if encounter new event
+let decodedEvent: Record<string, DecodedEvent> = {};
+//only single message will be stored, reset if encounter new message
+let decodedMessage: Record<string, DecodedMessage> = {};
```
Comment (stwiname) [inline packages/substrate-wasm/src/index.ts:159]: Do you think its an issue that items never get removed from these and would continuously grow?

**Review by stwiname — APPROVED** [review]

**Comment by stwiname** [comment]
Do we need updates to github actions so that it publishes?

**Pull request merged**